cpp/src/gandiva/precompiled/string_ops.cc

```diff
@@ -191,38 +191,73 @@ VAR_LEN_TYPES(IS_NOT_NULL, isnotnull)
 #undef IS_NOT_NULL
 
 FORCE_INLINE
-const char* substr_utf8_int64_int64(int64 context, const char* input, int32 in_len,
-                                    int64 offset64, int64 length, int32* out_len) {
-  if (length <= 0 || input == nullptr || in_len <= 0) {
-    *out_len = 0;
+const char* substr_utf8_int64_int64(int64 context, const char* input, int32 in_data_len,
+                                    int64 offset64, int64 length, int32* out_data_len) {
+  if (length <= 0 || input == nullptr || in_data_len <= 0) {
+    *out_data_len = 0;
+    return "";
+  }
+
+  int32 in_glyphs_count = utf8_length(context, input, in_data_len);
+
+  // handle invalid glyphs in input
+  if (in_glyphs_count == 0) {
+    *out_data_len = 0;
     return "";
   }
 
   int32 offset = static_cast<int32>(offset64);
-  int32 startIndex = offset - 1;  // offset is 1 for first char
+  int32 from_glyph = offset - 1;  // offset is 1 for first char
   if (offset < 0) {
-    startIndex = in_len + offset;
+    from_glyph = in_glyphs_count + offset;
   } else if (offset == 0) {
-    startIndex = 0;
+    from_glyph = 0;
   }
 
-  if (startIndex < 0 || startIndex >= in_len) {
-    *out_len = 0;
+  if (from_glyph < 0 || from_glyph >= in_glyphs_count) {
+    *out_data_len = 0;
     return "";
   }
 
-  *out_len = static_cast<int32>(length);
-  if (length > in_len - startIndex) {
-    *out_len = in_len - startIndex;
+  int32 out_glyphs_count = static_cast<int32>(length);
+  if (length > in_glyphs_count - from_glyph) {
+    out_glyphs_count = in_glyphs_count - from_glyph;
   }
 
-  char* ret = reinterpret_cast<char*>(gdv_fn_context_arena_malloc(context, *out_len));
+  int32 start_pos = 0;
+  int32 end_pos = in_data_len;
+
+  int32 n_glyphs = 0;
+  int32 pos = 0;
+  while (pos < static_cast<int64_t>(in_data_len)) {
+    if (n_glyphs == from_glyph) {
+      start_pos = pos;
+    }
+    int32 glyph_len = utf8_char_length(input[pos]);
+    // Skip invalid glyphs.
+    if (glyph_len == 0) {
+      glyph_len = 1;
+    }
+    pos += glyph_len;
+    if (n_glyphs - from_glyph + 1 == out_glyphs_count) {
+      end_pos = pos;
+    }
+    n_glyphs++;
+  }
+
+  if (end_pos > in_data_len || end_pos < 0) {  // <0 check to handle 'pos' overflow
```

Comment:
Please use int64 instead so that overflow can be checked as:
`end_pos > numeric_limits<int32>::max()`. When might end_pos be greater than in_data_len?

Reply:
end_pos can be greater than in_data_len when the last byte in the input is the start of a multibyte character